src/TypeSpec.Extension/Emitter.Csharp/src/type/converter.ts

```diff
@@ -36,22 +229,381 @@ export function fromSdkEnumType(
+        case "uri":
+        case "url":
+            return InputPrimitiveTypeKind.Uri;
+        case "boolean":
+            return InputPrimitiveTypeKind.Boolean;
+        case "plainDate":
+            return InputPrimitiveTypeKind.Date;
+        case "plainTime":
+            return InputPrimitiveTypeKind.Time;
+        case "any":
+            return InputPrimitiveTypeKind.Object;
+        case "int16": // TODO: add support for those types
+        case "uint16":
+        case "uint32":
+        case "uint64":
+        case "ipV4Address":
+        case "ipV6Address":
+        case "password":
+        case "armId":
+        case "ipAddress":
+        default:
+            throw new Error(`Unsupported built-in type kind '${type.kind}'`);
+    }
+}
+
+function fromScalarType(scalarType: SdkType): InputPrimitiveType {
+    return {
+        Kind: InputTypeKind.Primitive,
+        Name: getCSharpInputTypeKindByPrimitiveModelName(
+            scalarType.kind,
+            scalarType.kind,
+            undefined // To-DO: encode not compatible
+        ),
+        IsNullable: scalarType.nullable
+    };
+}
+
+function fromIntrinsicType(scalarType: SdkType): InputIntrinsicType {
+    const name = (scalarType.__raw! as IntrinsicType).name;
+    return {
+        Kind: InputTypeKind.Intrinsic,
+        Name: getCSharpInputTypeKindByIntrinsic(
+            scalarType.__raw! as IntrinsicType
+        ),
+        IsNullable: false
+    };
+}
+
+export function fromUnionType(
+    union: SdkUnionType,
+    context: SdkContext,
+    models: Map<string, InputModelType>,
+    enums: Map<string, InputEnumType>
+): InputUnionType | InputType {
+    let itemTypes: InputType[] = [];
+    for (const value of union.values) {
+        const inputType = fromSdkType(value, context, models, enums);
+        itemTypes.push(inputType);
+    }
+
+    return itemTypes.length > 1
+        ? {
+              Kind: InputTypeKind.Union,
+              Name: InputTypeKind.Union,
+              UnionItemTypes: itemTypes,
+              IsNullable: false
+          }
+        : itemTypes[0];
+}
+
+function fromSdkConstantType(
+    constantType: SdkConstantType,
+    enums: Map<string, InputEnumType>,
+    literalTypeContext?: LiteralTypeContext
+): InputLiteralType {
+    return {
+        Kind: InputTypeKind.Literal,
+        Name: InputTypeKind.Literal,
+        LiteralValueType:
+            constantType.valueType.kind === "boolean" ||
+            literalTypeContext === undefined
+                ? fromSdkBuiltInType(constantType.valueType)
+                : convertConstantToEnum(
+                      constantType,
+                      enums,
+                      literalTypeContext
+                  ),
+        Value: constantType.value,
+        IsNullable: false
+    };
+
+    function convertConstantToEnum(
+        constantType: SdkConstantType,
+        enums: Map<string, InputEnumType>,
+        literalTypeContext: LiteralTypeContext
+    ) {
```

Comment:
maybe we also need a TODO here because we will not be doing this in a near future